xarray/core/dataset.py

```
@@ -1095,6 +1095,88 @@ def maybe_chunk(name, var, chunks):
+        indexers_list = []
+        for k, v in iteritems(indexers):
+            if isinstance(v, integer_types + (slice, Variable)):
+                pass
+            elif isinstance(v, DataArray):
+                v = v.variable
+            elif isinstance(v, tuple):
+                v = as_variable(v)
+            elif isinstance(v, Dataset):
+                raise TypeError('cannot use a Dataset as an indexer')
+            else:
+                v = np.asarray(v)
+            indexers_list.append((k, v))
+        return indexers_list
+
+    def _get_indexers_coordinates(self, indexers):
+        """  Extract coordinates from indexers.
+        Returns an OrderedDict mapping from coordinate name to the
+        coordinate variable.
+
+        Coordinates to be extracted and attached should satisfy
+        + Dimension coordinate of the indexers.
+          Non-dimension coordinate of the indexers are not attached.
+        + Only coordinate with a name different from any of sef.variables.
+
+        If self already has the same name coordinate, we raise an ValueError.
+        """
+        from .dataarray import DataArray
+        import warnings
+
+        coord_list = []
+        for k, v in indexers.items():
+            if isinstance(v, DataArray):
+                coords = {d: v.coords[d].variable for d in v.dims
+                          if d in v.coords}
+                if v.dtype.kind == 'b' and v.dims[0] in coords:
+                    # Make sure in case of boolean DataArray, its
+                    # coordinate should be also indexed.
+                    assert v.ndim == 1  # we only support 1-d boolean array
+                    coords[v.dims[0]] = coords[v.dims[0]][v.variable]
+
+                for k, vc in self.variables.items():
+                    if k in coords and not vc[v.values].equals(coords[k]):
+                        # TODO raise an Error in the next release
+                        warnings.warn(
```

Errorning coordinate conflict may result in significant backward incompatibility. 
I just warn `FutureWarning` here if there is any conflict, but I believe it will raise an Error in the future version.

---

I'm not convinced we need to error here at all. In methods like arithmetic, we silently drop conflicting non-index coordinates.